sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PullOutJoinCondition.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, PredicateHelper}
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.JOIN
+
+/**
+ * This rule ensures that [[Join]] condition doesn't contain complex expressions in the
+ * optimization phase.
+ *
+ * Complex condition expressions are pulled out to a [[Project]] node under [[Join]] and are
+ * referenced in join condition.
+ *
+ * {{{
+ *   SELECT * FROM t1 JOIN t2 ON t1.a + 10 = t2.x ==>
+ *   Project [a#0, b#1, x#2, y#3]
+ *   +- Join Inner, ((spark_catalog.default.t1.a + 10)#8 = x#2)
+ *      :- Project [a#0, b#1, (a#0 + 10) AS (spark_catalog.default.t1.a + 10)#8]
+ *      :  +- Filter isnotnull((a#0 + 10))
+ *      :     +- Relation default.t1[a#0,b#1] parquet
+ *      +- Filter isnotnull(x#2)
+ *         +- Relation default.t2[x#2,y#3] parquet
+ * }}}
+ */
+object PullOutJoinCondition extends Rule[LogicalPlan]
+  with JoinSelectionHelper with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(_.containsPattern(JOIN)) {
+    case j @ Join(left, right, _, Some(condition), _)
+        if j.resolved && !canPlanAsBroadcastHashJoin(j, conf) =>
```

Comment:
Do not pull out join condition if is is `BroadcastHashJoin`:
1. `BroadcastHashJoin` will not cause data skew.
2. `BroadcastHashJoin` condition will only be evaluated once.

Reply:
Can we put it as a code comment?

Reply:
Sorry, DSv2 does not support computeStats:
```
BUG: computeStats called before pushdown on DSv2 relation: parquet file:/private/var/folders/tg/f5mz46090wg7swzgdc69f8q03965_0/T/spark-a6b41d66-6342-4167-9af7-5a6b3a354abd/departments
java.lang.IllegalStateException: BUG: computeStats called before pushdown on DSv2 relation: parquet file:/private/var/folders/tg/f5mz46090wg7swzgdc69f8q03965_0/T/spark-a6b41d66-6342-4167-9af7-5a6b3a354abd/departments
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation.computeStats(DataSourceV2Relation.scala:80)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.default(SizeInBytesOnlyStatsPlanVisitor.scala:55)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.default(SizeInBytesOnlyStatsPlanVisitor.scala:27)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlanVisitor.visit(LogicalPlanVisitor.scala:47)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlanVisitor.visit$(LogicalPlanVisitor.scala:25)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.SizeInBytesOnlyStatsPlanVisitor$.visit(SizeInBytesOnlyStatsPlanVisitor.scala:27)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.$anonfun$stats$1(LogicalPlanStats.scala:37)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.stats(LogicalPlanStats.scala:33)
	at org.apache.spark.sql.catalyst.plans.logical.statsEstimation.LogicalPlanStats.stats$(LogicalPlanStats.scala:33)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.stats(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.optimizer.JoinSelectionHelper.canBroadcastBySize(joins.scala:305)
	at org.apache.spark.sql.catalyst.optimizer.JoinSelectionHelper.canBroadcastBySize$(joins.scala:304)
	at org.apache.spark.sql.catalyst.optimizer.PullOutJoinCondition$.canBroadcastBySize(PullOutJoinCondition.scala:46)
	at org.apache.spark.sql.catalyst.optimizer.JoinSelectionHelper.getBroadcastBuildSide(joins.scala:251)
	at org.apache.spark.sql.catalyst.optimizer.JoinSelectionHelper.getBroadcastBuildSide$(joins.scala:241)
	at org.apache.spark.sql.catalyst.optimizer.PullOutJoinCondition$.getBroadcastBuildSide(PullOutJoinCondition.scala:46)
```

Reply:
This is different from `PullOutGroupingExpressions`, as complex join keys do not affect other optimizer rules. Can we run this rule at the end of the optimizer?

Reply:
Can we put it in `Infer Filters`? This can infer more filters mentioned in https://github.com/apache/spark/pull/28642:
https://github.com/apache/spark/blob/7f6edf8118b39a5b1854c2d0a1fb9bbbb3c11f86/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L4231-L4254
![image](https://user-images.githubusercontent.com/5399861/140438871-efd80cb8-e348-454e-b561-e9cb27d71984.png)

Reply:
Sorry. If we need stats, `PullOutGroupingExpressions` should be after `Early Filter and Projection Push-Down`:
https://github.com/apache/spark/blob/36b3bbc0aa9f9c39677960cd93f32988c7d7aaca/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L211-L214